### ISSUE TITLE
How to deal with custom styles in dark mode?

### ISSUE THREAD

**Issue by xfq** [w3c/respec#4687]
## Description of problem

**URL to affected spec or repo:**

* Code: https://github.com/w3c/bp-i18n-specdev/pull/130
* Preview: https://deploy-preview-130--bp-i18n-specdev.netlify.app/

**What happened (e.g., it crashed)?:**

1. Open https://deploy-preview-130--bp-i18n-specdev.netlify.app/#lang_misc when the system is in light mode.
2. Select 'dark' in the lower left corner of the spec.
3. The colour is incorrect.

If the OS is in dark mode, the colour is correct.

**Expected behavior (e.g., it shouldn't crash):**

Colours should follow the `@media (prefers-color-scheme: dark)` media query.

**Comment by marcoscaceres** [comment]
I think the problem is with .links being applied to the `aside`. 

So that local.css should define 

`@media (prefers-color-scheme: dark) {
    .links { /**/}
}`

**Comment by sidvishnoi** [comment]
Using the media query won't work with forced dark-mode via the toggle in sidebar (that comes from fixup.js). 

So, the best strategy would be to use one of the variables from https://www.w3.org/StyleSheets/TR/2021/dark.css (and https://www.w3.org/StyleSheets/TR/2021/base.css) to support consistent dark mode styles (including custom styles). Even  with ReSpec's dark mode support, I avoided media query as much as possible and used variables from W3C stylesheet.

**Comment by sidvishnoi** [comment]
Opinion: I'm not in favor of existence of toggles. The `prefers-color-scheme` is already the preference set by user that we're after. Adding site-wise color scheme should be something either browsers or extensions should do.

**Comment by sidvishnoi** [comment]
Alternative would be to watch whether dark.css applies using JS (mutation observer on that link element), and then add a class to `html` and use that instead of (only) media query. But yeah, mostly the pain of having a toggle.

**Comment by marcoscaceres** [comment]
I like/use the toggle myself... because I don't like the specs on the dark background, even when the rest of the OS is in dark mode... so I get it.

**Comment by sidvishnoi** [comment]
So maybe fixup.js should add a `.dark` class when it's using dark mode. Then we can adapt the styles on our end - using media query along with :not(.dark).
Can you file an issue there (tr-design)?

**Comment by xfq** [comment]
I see. Thanks.

Filed https://github.com/w3c/tr-design/issues/349

**Comment by sidvishnoi** [comment]
Someday in future we can use [style queries](https://developer.mozilla.org/en-US/docs/Web/CSS/@container#container_style_queries) to simplify that. Someday...

**Comment by xfq** [comment]
fixup.js now adds a `.darkmode` class when it's using dark mode.

**Comment by tabatkins** [comment]
The better route, rather than add a class to the body, is to add (or manipulate an existing) `<meta name=color-scheme>` to have `content="light"` or `content="dark"`, which'll force the page into light/dark mode and make the MQ match appropriately.

**Comment by tabatkins** [comment]
(And ideally, generate the spec with `content="light dark"` so it'll respect the OS setting by default.)

**Comment by marcoscaceres** [comment]
Hmmm... so, we do generate `content="light dark"`,  but we should be checking if the author has included a `<meta name=color-scheme>` in the document first. 

I think we should:

* Check if `meta name=color-scheme` is present. If no, add it with "light dark".
* Deprecate and remove `conf.darkMode`

That would solve for this.

**Comment by marcoscaceres** [comment]
Sent https://github.com/w3c/respec/pull/4700

**Comment by tabatkins** [comment]
Sigh, I thought it did, but I guess I just tested its effect on the used color scheme, not the MQ. It looks like it does indeed not affect the MQ. I'm gonna raise this in the CSSWG.

**Comment by xfq** [comment]
Tracked in https://github.com/w3c/csswg-drafts/issues/10249

**Comment by xfq** [comment]
Any progress on this issue? I assume we are able to solve this in a way that does not depend on changes to CSS specs?

**Comment by iherman** [comment]
@xfq I had to fight a bit with dark mode on our specs myself. Here is the pattern I use in CSS:

```css
XYZ {
   // your extra styling for light mode
}
body.darkmode XYZ {
    // your extra styling for dark mode here
}
```

What happens is that the small javascript, that is triggered by the dark/light mode switch in the lower left-hand corner of the screen, modifies the class of the `<body>` element. The default css is in light mode and then the body element is classified to dark if this is your default.

I hope this helps.

**Comment by xfq** [comment]
Thanks @iherman! I tried it in https://github.com/w3c/bp-i18n-specdev/pull/130 and it seemed to work. My only problem is that when my OS was in light mode, I selected auto in the lower left corner of the screen, but for some reason, there is still the class `darkmode` on the body.

**Comment by iherman** [comment]
Yeah, I also had some problems in the past with that script, though they were solved. I guess that is not a respec issue, though, it should be reported somewhere else. @deniak knows where to submit these...

**Comment by niklasl** [comment]
See proposed fix for this in https://github.com/speced/respec/pull/4925. It relies on the CSS features `:has()` and `:not()`) to work even with JS disabled.

**Comment by marcoscaceres** [comment]
This was addressed by #4925 (merged 2026-04-12), which added `color-scheme: dark` support to the body element when the manual theme toggle is used.

For custom author styles that need to adapt to dark mode, the recommended pattern is:

```css
body:has(input[name='color-scheme'][value='dark']:checked) .your-selector {
  /* dark mode overrides */
}
```

Alternatively, the upstream `fixup.js` adds a `.darkmode` class to the body, so:

```css
body.darkmode .your-selector {
  /* dark mode overrides */
}
```

For OS-level dark mode (without the manual toggle), use `@media (prefers-color-scheme: dark)` as usual.

Closing as the ReSpec-side fix is in place. The remaining work is documentation for spec authors.

**Comment by tabatkins** [comment]
And note that the CSSWG resolved at our last f2f to finally fix this properly, so `<meta name=color-scheme>` *does* affect the `(prefers-color-scheme)` MQ, so all your styles can just rely on the MQ even if you have a manual theme switcher.

Not implemented yet, but things are in motion!

**Comment by marcoscaceres** [comment]
Oh nice! Found it: https://bugs.webkit.org/show_bug.cgi?id=312205
